pagefaultgames/pokerogue
# [Bug] Fix Wimp Out/EE Trainer Battle causing enemy stack with multi hit

## What are the changes the user will see?
Activating Wimp Out or Emergency Exit on a boss pokemon (Guzma 115) will no longer cause the ability to trigger twice and cause 2 pokemon to stack on top of one another

## Why am I making these changes?
Bug Thread: https://discord.com/channels/1125469663833370665/1290778513363959941

## What are the changes from a developer perspective?


## Screenshots/Videos


## How to test the changes?
Basically this issue started happening due to Wimp Out and Emergency Exit triggering twice due to the interaction with the boss break bars. One for Multi-hit, which waits until the end of damage, and the other is just for general damage. 

To prevent this I've added more guardrails around the applyAbAttrs to prevent this from happening again. Now there should be no overlap between the two applyAbAttrs.

I added an optional argument `move` to `damageAndUpdate` in `pokemon.ts`. This allows me to check for `MultiHitAttr`

I moved the multi-hit check to `move-effect-phase.ts` to utilize the `lastHit` variable. This will make the effect more consistent rather than relying on hitCount or hitsLeft.

### Videos

## Before
<details>
  <summary>Multi-Hit Move</summary>

https://github.com/user-attachments/assets/c717072f-c72f-4bd3-8e50-aa626ed1bf58

</details>
<details>
  <summary>Multi-Lens</summary>

https://github.com/user-attachments/assets/963d77db-20a9-4421-8c62-6b89e61ea81e

</details>
<details>
  <summary>Parental Bond</summary>

https://github.com/user-attachments/assets/f7bd501f-b651-44c1-936a-f3dfcd949fb8

</details>

## After
<details>
  <summary>Multi-Hit Move</summary>

https://github.com/user-attachments/assets/53ca41e9-6029-46ed-9149-4392ec48013c

</details>
<details>
  <summary>Multi-Lens</summary>

https://github.com/user-attachments/assets/ec04cba4-c871-4dff-9cc1-3d5d51759fe2

</details>
<details>
  <summary>Parental Bond</summary>

https://github.com/user-attachments/assets/19f770c6-21ff-44bf-993c-997af4547f87

</details>

## Checklist

Are there any localization additions or changes? If so:
